rtic-sync/src/signal.rs

```diff
@@ -0,0 +1,119 @@
+//! A "latest only" value store with unlimited writers and async waiting.
+
+use core::{
+    cell::UnsafeCell,
+    future::poll_fn,
+    sync::atomic::{fence, Ordering},
+    task::Poll,
+};
+use rtic_common::waker_registration::CriticalSectionWakerRegistration;
+
+/// Basically an Option but for indicating
+/// whether the store has been set or not
+#[derive(Clone, Copy)]
+enum Store<T> {
+    Set(T),
+    Unset,
+}
+
+/// A "latest only" value store with unlimited writers and async waiting.
+pub struct Signal<T: Copy> {
+    waker: CriticalSectionWakerRegistration,
+    store: UnsafeCell<Store<T>>,
+}
+
+unsafe impl<T: Copy> Send for Signal<T> {}
+unsafe impl<T: Copy> Sync for Signal<T> {}
+
+impl<T: Copy> Signal<T> {
+    /// Create a new signal.
+    pub const fn new() -> Self {
+        Self {
+            waker: CriticalSectionWakerRegistration::new(),
+            store: UnsafeCell::new(Store::Unset),
+        }
+    }
+
+    /// Split the signal into a writer and reader.
+    pub fn split(&'static self) -> (SignalWriter<T>, SignalReader<T>) {
+        (SignalWriter { parent: self }, SignalReader { parent: self })
+    }
+}
+
+/// Fascilitates the writing of values to a Signal.
+#[derive(Clone)]
+pub struct SignalWriter<T: 'static + Copy> {
+    parent: &'static Signal<T>,
+}
+
+impl<T: Copy> SignalWriter<T> {
+    /// Write a raw Store value to the Signal.
+    fn write_inner(&mut self, value: Store<T>) {
+        fence(Ordering::SeqCst);
```

Comment:
This fence looks like it's not needed, what's the reason for it?

Reply:
To be honest I don't understand compiler fencing and simply included it because during my research looking through the other rtic-sync structures I saw it used just before critical sections often. I'll remove them.

Reply:
In this case the critical section does everything with the shared state so there should not be any need for extra synchronization :rocket: